File: testcases/validparenthesis/validparenthesis.lua
import("array")

function valid_paren(s)
  paren_table = table()
  paren_table["("] = ")"
  paren_table["["] = "]"
  paren_table["{"] = "}"
  stack = {}
  for i=0, #s do
    if paren_table[s[i]] ~= null then
      array.append(stack, s[i])
    else 
      element = array.pop(stack)
      if element == null then
        return false
      end
      if paren_table[element] ~= s[i] then
        return false
      end
    end
  end
  return #stack == 0
end

return valid_paren

File: testcases/validparenthesis/validparenthesis.js
const testCase = require('../test_fn.js');
var isValid = function (s) {
  const table = {
    '{': '}',
    '[': ']',
    '(': ')',
  };

  const stack = [];
  for (let i = 0; i < s.length; i++) {
    if (s[i] in table) {
      stack.push(s[i]);
      continue;
    }
    const el = stack.pop();
    if (!el) {
      return false;
    }
    if (table[el] !== s[i]) {
      return false;
    }
  }
  return stack.length <= 0;
};

testCase(['()()(())'], isValid);
testCase(['()()(())((('], isValid);
testCase(['()()(())((([{[]}])))[][][][][]()()((()))'], isValid);
testCase(['()()(())((([{[]}])))[][][][][]()()((())'], isValid);
testCase(['()()(())((([{[]}])))][][][][]()()((())'], isValid);
testCase(['(())[()(())((([{[]}])))[][][][][]()()((()))]'], isValid);
